README
*** System Configurator Project ***

System Configurator is a project which attempts to be an automated system
setup utility for Linux.

It smooths over differences between Linux distributions and hardware platforms
by providing a unified API to set up networking, hardware, and boot strapping.

System Configurator is heavily developed by a 
crack team of ring tailed lemurs.  Any issues should be directed to their feeder
at sisuite-users@lists.sourceforge.net.

It is still in its infancy, so if you wish to help out, please visit the 
System Configurator home page (http://systemconfig.sourceforge.net).

*** NOTE ***

System Configurator is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License located in the COPYING file of this
distribution for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

*** NOTE 2 ***

No animals, real, imaginary, or fantastic were harmed during the creation
of this software.

*** NOTE 3 ***

There is a possible exception to NOTE 2, as significant quantities of sushi
were consumed during the design of this project.

*** NOTE 4 ***

To build the RPM package without being root, execute the following command:
    rpmbuild -ba --define '%check exit 0' sc.spec


CHANGELOG
- 2.2.11 Sat 14 2013
	* Migrated to INSTALLDIRS=vendor in SystemConfig module.
- 2.2.2 Mar 26 2005
	* added kernel version detection patch from Erik Jacobson
	* updated hardware detection module
	* support modprobe.conf for 2.6 kernels
	* 2.2.0 & 2.2.1 never released as they were test cuts for 
          these features

- 2.0.10 Mar 14, 2004
	* add EFI path used by SLES9, patch from Erik Jacobson
	* Fix path for mkinitrd on Debian systems, patch from vvidic

- 2.0.9 Nov 19, 2003
	* Alpha support for, er alpha (aboot, to be precise)
	* Some fixes for EFI style elilo booting, including knowledge
	  of where sgi stores their elilo.efi.
	* add xfs module mapping to Root.pm (thanks to Erik Jacobson)

- 2.0.8 Jul 10, 2003
	* Fixed bug in hardware ordering which was missed as common
	  case worked properly.

- 2.0.7 Apr 21, 2003
	* Changes to Devfs.pm for device transformation to work with 
	  cciss and ida devices..

- 2.0.6 Apr 11, 2003
        * make file handle local to fix grub problems during deployment.

- 2.0.5 Jan 21, 2003
        * Added [NETWORK] SHORTHOSTNAME (boolean) to allow for short
          host names to be used during install (SuSE8 and LinuxConf
          modules (i.e. RedHat and Mandrake) now support this
        * Added [BOOT] PREFERRED to allow for overriding the preferred
          order to try to setup boot loaders.  The only use at this
          point is to specify "Boot::Grub" to get grub used on a
          system that has both grub and lilo functional
        * Do the right thing is a domain name is not specified in
          SuSE8 module

- 2.0.4 Jan 9, 2003
	* Fixed bug in Time::Zone modules, now they actually work :)
	  (previously they would barf on filesmod call)
	* Allow for '.' in release number for kernel (to allow RedHat's
	  2.4.18-19.7.x notation)
	* Changed hardware behavior.  Group of modules happens now
	  as well as the ability to order via [HARDWARE] ORDER = XX YY

- 2.0.3 Nov 8, 2002
	* added iSeries Hardware detection
	* UL 1.0 & SLES8 adjustments to SuSE8 networking module
	* Final YaBoot fixes (tested on Power3 SuSE 7.1 and SLES8)
	* Removed man 3 pages from rpm package

- 2.0.2 Oct 22, 2002
	* Fix for SuSE initrd generation, hardware modules
	  now in the right order, and exit 9 is ignored.

- 2.0.1 Oct 16, 2002
	* Hardware refresh
	* added hardware.lst override cabability
	* grub finally works in System Imager 3.0 (devfs translation
	  works now)
	* updated systemconfigurator man page

- 2.0.0 Oct 1, 2002
	* Final doc fixes
	* Grub support fixed

- 1.9.6 Sept 10, 2002
	* iSeries fixes

- 1.9.5 July 29, 2002
	* Redid all config parsing code (less warnings about undefined
	  variables should now exist)
	* Added more config parsing tests

- 1.9.4 June 12, 2002
	* Fixed a number of SuSE8 networking setup issues.
	* Made Time Zone code actually work.
	* Created a hack to reset KEYTABLE on Suse 7.x to default (US).  
	  It was defaulting to german quertz keyboard without this.
	* Made it possible to turn off timezone, keyboard, and userexit portions of code

- 1.9.3 June 11, 2002
        Minor Bug Update
        Bugs Fixed:
                * Module detection doesn't blow up on ext2 root
                * Initrd builds properly on SuSE during install
	Features:
		* Added Debian Time Zone support

- 1.9.2 June 11, 2002
        Major Feature Update
        Features:
                * Finished SuSE kernel module detection
                * Ramdisks now built for SuSE 7.x and 8.0
                * TimeZone config files built for Red Hat (and cousins) and SuSE
                           (Debian still coming) 

- 1.9.1 June 5, 2002
	First development release on road to 2.0.0 (sorry for changing
	numbering schemes again)
	Major Feature Update
	Features:
		* Added UserExit capabilities
		* Added SuSE 8.x network support
		* Added SuSE kernel module detection (incomplete)
		* Hardware Table refresh
		
	Known Bugs:
		* Ramdisks don't get built for SuSE at all yet 
		* Ramdisks don't get built for any distro without
		    a mkinitrd script
		* SuSEconfig is run way too many times (need a generic
		    way to run this at the end)
		* TimeZone config files aren't written out properly
		* Docs are out of date for things like UserExit
		* Grub still has some issues that need to be addressed for
		    Red Hat 7.3

	Use of this version of SC is only suggested for Red Hat, Mandrake,
	Conectiva and Debian users.  All other users should stay with the
	last stable version.

- 1.25  April 3, 2002 (This was only released into the OSCAR CVS tree)
	Minor Feature Update
	Features: 
		* Changed the way Initial Ramdisk generation was done
		  (not complete yet, but works for Debian and Red Hat)
		* Added Support for setting up Timezones (incomplete)
		
- 1.20  Feb 20, 2002
	Minor Feature Update
	Features:
		* Added support for gzipped modules in building
		  an initial ramdisk
		* Addition of VGA section in [BOOT]
		* Addition of EXTRAS section in [BOOT]. (See 
		  systemconfig.conf(5) for more details)
		* Update of the Hardware::PCI::Table module with
		  newer PCI ids
	Bugs:
		* made the tests run cleaner

- 1.10  Feb 12, 2002
	Minor bugfix release
	Bug Fixes:
		* Lilo updates

- 1.01  Dec 7, 2001
	Minor bugfix release
	Bug Fixes:
		* Elilo fixes for Debian

- 1.00  Criton (Oct 25, 2001)
	Minor bugfix release
	Bug Fixes:
		* EFI fixes for Red Hat
		* SuSE Yaboot fixes

- 0.88  Sun (Oct 1, 2001)
	Major feature release
	Features:
		* Addition of Elilo support (Debian IA64)
		* Addition of EFI support (Red Hat IA64)
	Bug Fixes:
		* cleanup of debian packages
		* cleanup of rpm packages
		* few more Grub fixes

- 0.87  D'Argo + 2 (Aug 21, 2001)
        Major feature release
        Features:
                * Addition of Palo support (HPA Risc bootloader)
                * Addition of Initial Ramdisk generating libraries
        
- 0.86  D'Argo + 1 (Aug 14, 2001)
        Minor bug fix release
        Bug Fixes:

lib/Initrd2.pm
package Initrd2;

#   $Id: Initrd2.pm 664 2007-01-15 21:40:36Z arighi $

#   Copyright (c) 2001-2005 International Business Machines

#   This program is free software; you can redistribute it and/or modify
#   it under the terms of the GNU General Public License as published by
#   the Free Software Foundation; either version 2 of the License, or
#   (at your option) any later version.
 
#   This program is distributed in the hope that it will be useful,
#   but WITHOUT ANY WARRANTY; without even the implied warranty of
#   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#   GNU General Public License for more details.
 
#   You should have received a copy of the GNU General Public License
#   along with this program; if not, write to the Free Software
#   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

#   Sean Dague <sean@dague.net>
#   Vasilios Hoffman <greekboy@users.sourceforge.net>

use strict;
use Carp;
use POSIX;
use File::Copy;
use Util::Log qw(:all);
use Util::Cmd qw(:all);
use base qw(Exporter);
use vars qw(@EXPORT $VERSION);

$VERSION = sprintf("%d", q$Revision: 664 $ =~ /(\d+)/);

=head1 NAME

Initrd - builds initial ramdisks for kernels

=head1 SYNOPSIS

Initrd::setup($config);

=head1 DESCRIPTION

The Initrd module generates initial ramdisks for kernels, provided there is enough userland tools and kernel support to allow this.

=head1 AUTHOR

  Sean Dague <sean@dague.net>
  Vasilios Hoffman <greekboy@users.sourceforge.net>
  Erich Focht <efocht@hpce.nec.com>

=head1 SEE ALSO

L<Config>, 
L<perl>

=cut

sub setup {

    my $config = shift;

    my %kernels = $config->varlist("^kernel");
    my $root = $config->get("root");

    my (@ramdisks,@files) = ();
    my $ramdisk;

    foreach my $key (sort keys %kernels) {
        if($key =~ /(kernel\d+)_path/) {
            my $ramdisk = $1 . "_initrd";
            my $kernel = $kernels{$1 . "_path"};
            if (! exists $kernels{$ramdisk}) {
                debug("configuring ramdisk for $1");
                eval {
                    my $rdfile = create_initrd($root, $kernel);
                    if($rdfile) {
                        $config->set("$ramdisk",$rdfile);
                    } else {
                        croak("couldn't create ramdisk for $1");
                    }
                };
                if($@) {
                    carp ("Ramdisk creation for kernel $kernel has failed");
                }
            }
        }
    }
}

sub create_initrd {
    my ($root, $kernel) = @_;
    my $version = kernel_version("$root/$kernel");
    my $rdfile = "/boot/sc-initrd-$version.gz";
    if( (uname)[4] eq "ia64") {
        # Itaniums need things in their special place
        $rdfile = "/boot/efi/sc-initrd-$version.gz";
    }
    my $outfile = $root . $rdfile;
    if( -e "$root/etc/mkinitrd/mkinitrd.conf" ) {
        # This is now Debian 3.0
        if(exec_mkinitrd_debian($root, $outfile, $version)) {
            return $rdfile;
        }
    } else {
        if(exec_mkinitrd($root, $outfile, $version)) {
            return $rdfile;
        }
    }
    return undef;
}

sub exec_mkinitrd {
    my ($root, $outfile, $version) = @_;
    my $cmd = "mkinitrd $root$outfile $version";
    return !system($cmd);
}

sub exec_mkinitrd_debian {
    my ($root, $outfile, $version) = @_;
    my $cmd = "mkinitrd -o $root$outfile /lib/modules/$version";
    return !system($cmd);

}

# Super simple kernel version.  Just open the kernel an
# look for a line like 2.4.8-27stuff in it 

sub kernel_version {
    my $file = shift;
    my $version = '0.0.0';

    if(_is_gzip($file)) {
        open(IN,"gzip -dc $file |") or croak("Couldn't run gzip -dc $file");
    } else {
        open(IN,"<$file") or croak("Failed to open $file.");
    }
        
    while(<IN>) {
        # When Linux Kernel 4.0pre1 comes out, we'll have to change this
        if(/Linux\ version\ ([123]\.\d+\.[\w\-\.]+)/) {
            verbose("Found version '$1' in line '$_'");
            $version = $1;
            last;
        }
    }
    if ($version eq "0.0.0") {
        # trying ia32 bzImage kind of thing
        my $buffer;
        if (seek(IN, 0x202, 0) && (read(IN, $buffer, 4) == 4) 
            && (substr($buffer,0,4) eq "HdrS")
            && seek(IN, 0x20e, 0) && (read(IN, $buffer, 2) == 2)) {
            my $ofs = 0x200 + unpack("S", substr($buffer,0,2));
            if (seek(IN, $ofs, 0) && (read(IN, $buffer, 80) == 80)) {
                $version = substr($buffer, 0, index($buffer, " "));
            }
        }
    }
    close(IN); 

    return $version;

}

sub _is_gzip {
    my $file = shift;
    debug("opening $file to figure out if it is gzip archive");
    open(IN,"<$file") or (carp($!), return undef);
    my $chr1 = getc IN;
    my $chr2 = getc IN;
    debug(unpack("H*",$chr1));
    debug(unpack("H*",$chr2));
    close(IN) or (carp($!), return undef);
    if(unpack("H*",$chr2) eq "8b" and unpack("H*",$chr1) eq "1f") {
        debug("Is a gzip archive");
        return 1;
    } else {
        debug("Is not a gzip archive");
        return 0;
    }
}

1;





